### DISCUSSION TITLE
PDF page cropping with respect to both visual part and elements (svg, text, etc.)

### DISCUSSION THREAD

**Discussion by vkovenko** [pymupdf/PyMuPDF#2579] [Q&A] [answered]
Greetings. Let's say that I have a fitz.Rect object, which define the bounding box I want to crop out from PDF. How to use it to crop the PDF w.r.t all the elements (svg, text, etc.)?

**Comment by JorjMcKie** [comment]
Geometry objects like rectangles, points and matrices exist without dependency to anything else in the world. They are nothing than the set of numbers by which they are defined.

The visible part of a PDF page is defined via its `CropBox`, a rectangle defined within the page's object definition inside the PDF.
PyMuPDF offers a simple way to set (change) that rectangle: `page.set_cropbox(rect)` where "rect" is some `fitz.Rect`.
To extract the current cropbox, simply look at `page.cropbox`.

Setting the cropbox only restricts what is visible of the page. The page content is not changed in any way if the cropbox is changed. Changing the cropbox again (e.g. setting it to the full `MediaBox`) will make other parts (or the full page) visible again.

**Comment by vkovenko** [comment]
@JorjMcKie , thanks for your answer! But is it possible in anyway to remove objects outside of cropbox in an elegant and easy way using fitz?

**Reply by JorjMcKie** [reply, marked as answer]
Yes. This works via so-called redactions. You define redaction annotations for desired rectangles and then call a page method to "apply" all redactions defined on it.
Thereafter, text content intersecting an annotation rectangle wil be physically gone. Same is true fro any hyperlinks. for images, there is a 3-fold choice:
1. Ignore them (leave untouched)
2. Delete if interscting a redaction
3. Punch a hole for every redaction annotation intersecting.

So to delete page content outside a given rectangle, you need to define (up to) four annotation rectangles covering the areas outside and then execute `apply_redactions()`.

**Reply by vkovenko** [reply]
Thank you very much for your answer!

**Reply by JorjMcKie** [reply]
Please do not hesitate to ask if more questions come up.

**Reply by JorjMcKie** [reply]
One more comment:
We are not able to remove vector graphics in this way currently. So there may be situations where awkward lines or similar are still visible.

**Reply by vkovenko** [reply]
That's okay, in practice, I delete such graphical elements by their visibility on image/pdf (in terms of pixels). Thank you one more time!

**Reply by JorjMcKie** [reply]
True images are no problem as I wrote, only vector graphics.

**Reply by StaleLoafOfBread** [reply]
@JorjMcKie Is there a recommended method to delete vector graphics? My end goal is to destructively crop a page.

**Reply by JorjMcKie** [reply]
Deletion of vector graphics is not possible currently. There are plans however to enable this via redactions - much like with images, but there is no defined schedule yet.
